src/shogun/kernel/SpectralMixtureKernel.cpp

```diff
@@ -0,0 +1,221 @@
+
+	return result;
+}
+
+SGVector<float64_t> CSpectralMixtureKernel::get_squared_features(CDotFeatures* features)
+{
+	SGVector<float64_t> sqr_features = SGVector<float64_t>(features->get_num_vectors());
+	
+	for (int i=0; i<sqr_features.size(); i++)
+		sqr_features[i] = features->dot(i, features, i);
+
+	return sqr_features;
+}
+
+SGMatrix<float64_t> CSpectralMixtureKernel::get_square_distance_matrix()
+{
+	CDotFeatures* dot_lhs = dynamic_cast<CDotFeatures*>(lhs);
+	CDotFeatures* dot_rhs = dynamic_cast<CDotFeatures*>(rhs);
+
+	REQUIRE(dot_lhs != NULL, "Left-hand-side features must be of type CDotFeatures");
+	REQUIRE(dot_rhs != NULL, "Right-hand-side features must be of type CDotFeatures");
+	
+	SGMatrix<float64_t> lhs_mtrx = dot_lhs->get_computed_dot_feature_matrix();
+	SGMatrix<float64_t> rhs_mtrx = dot_rhs->get_computed_dot_feature_matrix();
+
+	/* rhs^T . lhs: index (i, j) is dot product of lhs_i and rhs_j.
+	   result scaled by minus two for distance computation */
+	SGMatrix<float64_t> dot_mtrx = SGMatrix<float64_t>::matrix_multiply(rhs_mtrx, lhs_mtrx, true, false, -2.0);
+
+	SGVector<float64_t> sqr_lhs = get_squared_features(dot_lhs);
+	SGVector<float64_t> sqr_rhs = get_squared_features(dot_rhs);
+
+	for (int i=0; i<dot_mtrx.num_cols; i++)
+	{
+		/* To each column vector is added a scalar: lhs_i^2
+		   and the vector [rhs_1^2 ... rhs_n^2]^T*/
+		float64_t* col_vector_ptr = dot_mtrx.get_column_vector(i);
+
+		SGVector<float64_t>::add_scalar(sqr_lhs[i], col_vector_ptr, dot_mtrx.num_rows);
+		SGVector<float64_t>::add(col_vector_ptr, 1.0, col_vector_ptr, 1.0, sqr_rhs, dot_mtrx.num_rows);
+	}
+	
+	return dot_mtrx;
+}
+
+void CSpectralMixtureKernel::compute_weight_derivative(SGMatrix<float64_t>& squared_distances, index_t index)
+{
+	for (int i=0; i<squared_distances.num_rows; i++)
+	{
+		for (int j=0; j<squared_distances.num_cols; j++)
+		{
+			// 2*w_i*exp(-2*pi^2*dist^2*sigma_i^2) * cos(2*pi*dist*mu_i)
+			float64_t sqr_dist = squared_distances(i,j);
+			float64_t exp_factor = CMath::exp(-2*CMath::pow(CMath::PI,2)*sqr_dist*CMath::pow(m_stddeviations[index],2));
+			float64_t cos_factor = CMath::cos(2*CMath::PI*CMath::sqrt(sqr_dist)*m_means[index]);
+			squared_distances(i, j) = 2*m_weights[index]*exp_factor*cos_factor;
+		}
+	}
+}
+
+void CSpectralMixtureKernel::compute_mean_derivative(SGMatrix<float64_t>& squared_distances, index_t index)
+{
+	for (int i=0; i<squared_distances.num_rows; i++)
+	{
+		for (int j=0; j<squared_distances.num_cols; j++)
+		{
+			// -2*pi*dist*mu_i^2*exp(-2*pi^2*dist^2*sigma_i^2)*sin(2*pi*dist*mu_i)
+			float64_t sqr_dist = squared_distances(i,j);
+			float64_t outer_factor = -2*CMath::PI*CMath::sqrt(sqr_dist)*CMath::pow(m_weights[index],2);
+			float64_t exp_factor = CMath::exp(-2*CMath::pow(CMath::PI, 2)*sqr_dist*CMath::pow(m_stddeviations[index], 2));
```

Comment:
@karlnapf okay so [this](https://github.com/lambday/shogun/commit/e37d3a734c72805b3f817a1736c2f8517e9f3498) works. Unit-tests in my machine
```
[lambday@lambday.iitb.ac.in build]$ ./tests/unit/shogun-unit-test --gtest_filter="Elementwise_Sin.*"
Note: Google Test filter = Elementwise_Sin.*
[==========] Running 14 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 14 tests from Elementwise_Sin
[ RUN      ] Elementwise_Sin.SGMatrix_NATIVE
[       OK ] Elementwise_Sin.SGMatrix_NATIVE (1 ms)
[ RUN      ] Elementwise_Sin.SGMatrix_NATIVE_inplace
[       OK ] Elementwise_Sin.SGMatrix_NATIVE_inplace (0 ms)
[ RUN      ] Elementwise_Sin.SGVector_NATIVE
[       OK ] Elementwise_Sin.SGVector_NATIVE (0 ms)
[ RUN      ] Elementwise_Sin.SGVector_NATIVE_inplace
[       OK ] Elementwise_Sin.SGVector_NATIVE_inplace (0 ms)
[ RUN      ] Elementwise_Sin.SGMatrix_NATIVE_complex128
[       OK ] Elementwise_Sin.SGMatrix_NATIVE_complex128 (0 ms)
[ RUN      ] Elementwise_Sin.SGMatrix_NATIVE_complex128_inplace
[       OK ] Elementwise_Sin.SGMatrix_NATIVE_complex128_inplace (0 ms)
[ RUN      ] Elementwise_Sin.SGVector_NATIVE_complex128
[       OK ] Elementwise_Sin.SGVector_NATIVE_complex128 (0 ms)
[ RUN      ] Elementwise_Sin.SGVector_NATIVE_complex128_inplace
[       OK ] Elementwise_Sin.SGVector_NATIVE_complex128_inplace (0 ms)
[ RUN      ] Elementwise_Sin.SGMatrix_EIGEN3
[       OK ] Elementwise_Sin.SGMatrix_EIGEN3 (0 ms)
[ RUN      ] Elementwise_Sin.SGVector_EIGEN3
[       OK ] Elementwise_Sin.SGVector_EIGEN3 (0 ms)
[ RUN      ] Elementwise_Sin.CGPUMatrix_VIENNACL
[       OK ] Elementwise_Sin.CGPUMatrix_VIENNACL (69 ms)
[ RUN      ] Elementwise_Sin.CGPUMatrix_VIENNACL_inplace
[       OK ] Elementwise_Sin.CGPUMatrix_VIENNACL_inplace (59 ms)
[ RUN      ] Elementwise_Sin.CGPUVector_VIENNACL
[       OK ] Elementwise_Sin.CGPUVector_VIENNACL (0 ms)
[ RUN      ] Elementwise_Sin.CGPUVector_VIENNACL_inplace
[       OK ] Elementwise_Sin.CGPUVector_VIENNACL_inplace (0 ms)
[----------] 14 tests from Elementwise_Sin (129 ms total)

[----------] Global test environment tear-down
[==========] 14 tests from 1 test case ran. (129 ms total)
[  PASSED  ] 14 tests.
```
Have to take care of a few beauty related stuffs and documentation so didn't send this yet as a PR. Few things I want to point about this design

- I wanted to turn off implicit type conversion between CPU/GPU types following our previous discussion. So for elementwise operations, I have turned off default backend. Devs **must** specify backend and use supported type to use these. For example, `SGMatrix` with `ViennaCL` not supported. In order to use ViennaCL, devs should transfer it to GPU first and then call the method on that (when that makes sense).
- As a result of the above, in-place operations are available for `NATIVE` and  `VIENNACL` (Eigen3's `array()` thing doesn't work in-place I'm afraid). Separate wrappers for in-place and regular operation. In-place will work when the scalar types agree, `static_assert` fail otherwise (I thought it is better to give them a verbose info about why it failed *during* the compilation).
- Matrices and vectors don't have to be treated differently. Made use of the templated type instead to avoid rewriting. Introduced standard `data()` and `size()` methods in our matrix/vector types to achieve that.
- Future elementwise operation addition would be copy-paste (macros maybe)

Preparing a benchmark now to see if we gain performance using GPU over CPU for these.